modules/natgw/variables.tf

```diff
@@ -29,84 +19,149 @@ variable "tags" {
+  public_ip = {
+    create              = false
+    name                = "existing-public-ip-name"
+    resource_group_name = "external-rg-name"
+  }
+  ```
   EOF
-  default     = false
-  type        = bool
+  default     = null
+  type = object({
+    create              = bool
+    name                = string
+    resource_group_name = optional(string)
+  })
 }
 
-variable "pip_prefix_length" {
+variable "public_ip_prefix" {
   description = <<-EOF
-  Number of bits of the Public IP Prefix. This basically specifies how many IP addresses are reserved. Azure default is `/28`.
-
-  This value can be between `0` and `31` but can be limited by limits set on Subscription level.
+  A map defining a Public IP Prefix resource.
+  
+  List of available properties:
+
+  - `create`              - (`bool`, required) controls whether a Public IP Prefix is created, sourced, or not used at all.
+  - `name`                - (`string`, required) name of a created or sourced Public IP Prefix.
+  - `resource_group_name` - (`string`, optional) name of a resource group hosting the sourced Public IP Prefix resource, ignored
+                            when `create = true`.
+  - `length`              - (`number`, optional, defaults to `28`) number of bits of the Public IP Prefix, this value can be
+                            between `0` and `31` but can be limited on subscription level (Azure default is `/28`).
+
+  The module operates in 3 modes, depending on combination of `create` and `name` properties:
+
+  `create` | `name` | operation
+  --- | --- | ---
+  `true` | `!null` | a Public IP Prefix resource is created in a resource group of the NAT Gateway
+  `false` | `!null` | a Public IP Prefix resource is sourced from a resource group of the NAT Gateway, the resource group can be
+                      overridden with `resource_group_name` property
+  `false` | `null` | a Public IP Prefix resource will not be created or sourced at all
+
+  Example:
+
+  ```hcl
+  # create a new Public IP Prefix, default prefix length is `/28`
+  public_ip_prefix = {
+    create = true
+    name   = "new-public-ip-prefix-name"
+  }
+
+  # source an existing Public IP Prefix from an external resource group
+  public_ip = {
+    create              = false
+    name                = "existing-public-ip-prefix-name"
+    resource_group_name = "external-rg-name"
+  }
+  ```
   EOF
   default     = null
-  type        = number
-}
-
-variable "existing_pip_prefix_name" {
-  description = "Name of an existing Public IP Prefix resource to associate with the NAT Gateway. Only for newly created resources."
-  default     = null
-  type        = string
-}
-
-variable "existing_pip_prefix_resource_group_name" {
-  description = "Name of a resource group hosting the Public IP Prefix resource specified in `existing_pip_name`. When omitted Resource Group specified in `resource_group_name` will be used."
-  default     = null
-  type        = string
+  type = object({
```

Comment:
maybe instead of adding a default here (it would be used only when there is no `public_ip_prefix`) add it to the `length` type definition?

Reply:
Which default you mean, `null` for entire object or `28` number for length property? I don't understand this remark.

Reply:
this code also changed since the comment